Mathlib/Tactic/LinearCombination.lean

```diff
@@ -60,51 +65,61 @@ using `+`/`-`/`*`/`/` on equations and values.
+    | .const c₁, .const c₂ => .const <$> ``($c₁ - $c₂)
+    | .proof p₁, .const c₂ => .proof <$> ``(pf_sub_c $p₁ $c₂)
+    | .const c₁, .proof p₂ => .proof <$> ``(c_sub_pf $p₂ $c₁)
+    | .proof p₁, .proof p₂ => .proof <$> ``(sub_pf $p₁ $p₂)
   | `(-$e) => do
-    match ← expandLinearCombo e with
-    | none => pure none
-    | some p => ``(neg_pf $p)
+    match ← expandLinearCombo ty e with
+    | .const c => .const <$> `(-$c)
+    | .proof p => .proof <$> ``(neg_pf $p)
   | `(← $e) => do
-    match ← expandLinearCombo e with
-    | none => pure none
-    | some p => ``(Eq.symm $p)
+    match ← expandLinearCombo ty e with
+    | .const c => return .const c
+    | .proof p => .proof <$> ``(Eq.symm $p)
   | `($e₁ * $e₂) => do
-    match ← expandLinearCombo e₁, ← expandLinearCombo e₂ with
-    | none, none => pure none
-    | some p₁, none => ``(pf_mul_c $p₁ $e₂)
-    | none, some p₂ => ``(c_mul_pf $p₂ $e₁)
-    | some p₁, some p₂ => ``(mul_pf $p₁ $p₂)
+    match ← expandLinearCombo ty e₁, ← expandLinearCombo ty e₂ with
+    | .const c₁, .const c₂ => .const <$> ``($c₁ * $c₂)
+    | .proof p₁, .const c₂ => .proof <$> ``(pf_mul_c $p₁ $c₂)
+    | .const c₁, .proof p₂ => .proof <$> ``(c_mul_pf $p₂ $c₁)
+    | .proof p₁, .proof p₂ => .proof <$> ``(mul_pf $p₁ $p₂)
   | `($e⁻¹) => do
-    match ← expandLinearCombo e with
-    | none => pure none
-    | some p => ``(inv_pf $p)
+    match ← expandLinearCombo ty e with
+    | .const c => .const <$> `($c⁻¹)
+    | .proof p => .proof <$> ``(inv_pf $p)
   | `($e₁ / $e₂) => do
-    match ← expandLinearCombo e₁, ← expandLinearCombo e₂ with
-    | none, none => pure none
-    | some p₁, none => ``(pf_div_c $p₁ $e₂)
-    | none, some p₂ => ``(c_div_pf $p₂ $e₁)
-    | some p₁, some p₂ => ``(div_pf $p₁ $p₂)
-  | e => do
-    let e ← elabTerm e none
-    let eType ← inferType e
-    let .true := (← withReducible do whnf eType).isEq | pure none
-    some <$> e.toSyntax
-  return result.map fun r => ⟨r.raw.setInfo (SourceInfo.fromRef stx true)⟩
+    match ← expandLinearCombo ty e₁, ← expandLinearCombo ty e₂ with
+    | .const c₁, .const c₂ => .const <$> ``($c₁ / $c₂)
+    | .proof p₁, .const c₂ => .proof <$> ``(pf_div_c $p₁ $c₂)
+    | .const c₁, .proof p₂ => .proof <$> ``(c_div_pf $p₂ $c₁)
+    | .proof p₁, .proof p₂ => .proof <$> ``(div_pf $p₁ $p₂)
+  | `($e ^ $n) => do
+    -- Fully pre-elaborate n to speed up exponentiation elaboration.
+    -- Assumption: mathlib only uses Pow instances that respect the default instances
+    -- for the exponent. For example, if the exponent is `37` it should be a `Nat`.
+    let n ← (withSynthesize <| elabTerm n none) >>= Expr.toSyntax
+    match ← expandLinearCombo ty e with
+    | .const c => .const <$> ``($c ^ $n)
+    | .proof p => .proof <$> ``(pow_pf $p $n)
+  | e =>
+    -- We have the expected type from the goal, so we can fully synthesize this leaf node.
+    withSynthesize do
+      let c ← withSynthesizeLight <| Term.elabTerm e ty
```

Comment:
In the "proof" case, the expected type being provided here as `ty` will be wrong -- that is, the type will actually be `(a : ty) = (b : ty)`, not `ty`.  Since it's still very fast, I assume this doesn't matter?

Reply:
One weird (but useful) thing about the `elabTerm` protocol is that the expected type is merely a suggestion. You have to use `elabTermEnsuringType` if you want to be sure the result has the given type.

This might in fact be a bad idea, but it seems to work. It definitely works if each proof is an fvar, but it gets less and less likely to work as the proof term gets more and more complicated (though perhaps those proofs had little hope of elaborating anyway when the expected type is none? and maybe we want do discourage complicated proofs in a `linear_combination` expression?)

Reply:
It occurred to me that it would be good to ensure the type like in `elabTermEnsuringType` in the `.const` case, just to give the proofs an easier time elaborating. This inserts coercions if necessary.

Reply:
Oops, ensuring the type changed how elaboration works in the Weierstrass file. I guess it wasn't a good idea.

Reply:
> One weird (but useful) thing about the `elabTerm` protocol is that the expected type is merely a suggestion.

Thanks for the explanation!